implement idiomatic errors

bhcleek opened the pull request:
Consider this a proposal to fix #22. I believe these changes represent a more idiomatic approach, but I welcome any feedback. Obviously, this breaks compatibility with previous releases. Once we reach an agreement on changes, I'll update README.md before we merge.

Return error types from exported functions

Change the Binder function type, Binder.Bind, and Validator.Validate to
have an error return value and modify their parameters to no longer
include the slice of errors.

Flatten field errors. Each error is associated with at most one field.

Stop exporting Error. Replace it with a new interface type, FieldError,
and rename Error to fieldError.

mholt commented on errors.go:
```diff
@@ -11,41 +11,46 @@ import (
+		// error can morph two ways: (1) it can indicate something
+		// wrong with the request as a whole, or (2) it can point to a
+		// specific problem with a particular input field.
+		fieldName string `json:"fieldNames,omitempty"`
```
When I created the binding library, it was the only one - and I think still is, from what I've seen - the only one that allows us to map an error to zero or more field names, which is a requirement that our company had when I was still working there. Is there a way we can preserve this feature? It definitely has its advantages.

bhcleek replied:
Yes, I can put that back. The reason I took it out was that there was nowhere in the package that set the `FieldNames` to a slice with more than one element, and any consumer of the package could easily group the field errors by classification to aggregate them if they wanted. But it's not a change I feel strongly about, and is arguably out of scope here.

mholt replied:
You're right on all counts. :+1: A user's custom Validate method could, however, make an error that spanned 0 fields or 2+ fields (which we did at my last company). I'll take a look at the rest of this after the holiday (state holiday, go figure!) - thanks for working on this!

bhcleek replied:
I'll make that change, and I may make Errors an interface, too. That would allow for more idiomatic checking of types, because if an interface is type asserted to a struct (e.g. `v, ok := err.(Errors)), then the value MUST be exactly that struct. As it stands, if `Errors` is embedded in a new type, then the new type can't be asserted to an `Errors` value.

bhcleek commented:
I reverted to supporting a slice of field names for each field error, and I also updated the documentation.

It now seems awkward (honestly, it's always seemed awkward to me, but now even more so) to have `Handle()` on `Errors`. What would you think about removing the `Handle` method in favor of either a separate `http.ResponseWriter` implementation that wraps `Errors` or else implementing `http.ResponseWriter` on `Errors`?

mholt commented:
I see your point.

Having written more web apps in Go than I had since I started this project, I wonder if error handling belongs in this package at all. After having [examined a few ways to do it](http://mwholt.blogspot.com/2015/05/handling-errors-in-http-handlers-in-go.html), I now see why it may not be ideal to hand over error handling to a strange function in some other package. There's nothing special in ours that anyone else couldn't do easily, and your app may want to handle it differently.

So while the Errors type still has value in the validation process, perhaps we don't need any HTTP error handler at all. What do you think we just nuke it?

bhcleek commented:
Agreed. It's gone.

bhcleek commented:
FWIW, the conflicts here are simply because of the recent changes to `Errors.Handle`. Since I removed it, there are conflicts, but they're obviously trivial to resolve.

mholt commented:
Forgive my delay - going to be busy for a while. I still need to look it over a little bit.

bhcleek commented:
No sweat. You're busy with Caddy!
